dbms/src/Storages/MergeTree/DiskSpaceMonitor.cpp

```diff
@@ -1,10 +1,223 @@
 #include <Storages/MergeTree/DiskSpaceMonitor.h>
 
+#include <Common/escapeForFileName.h>
+#include <Poco/File.h>
+
+/// @TODO_IGR ASK Does such function already exists?
+bool isAlphaNumeric(const std::string & s)
+{
+    for (auto c : s)
+        if (!isalnum(c) && c != '_')
+            return false;
+    return true;
+}
+
 namespace DB
 {
 
-UInt64 DiskSpaceMonitor::reserved_bytes;
-UInt64 DiskSpaceMonitor::reservation_count;
+std::map<String, DiskSpaceMonitor::DiskReserve> DiskSpaceMonitor::reserved;
 std::mutex DiskSpaceMonitor::mutex;
 
+DiskSelector::DiskSelector(const Poco::Util::AbstractConfiguration & config, const std::string & config_prefix, String default_path)
+{
+    Poco::Util::AbstractConfiguration::Keys keys;
+    config.keys(config_prefix, keys);
+
+    constexpr auto default_disk_name = "default";
+    bool has_default_disk = false;
+    for (const auto & disk_name : keys)
+    {
+        if (!isAlphaNumeric(disk_name))
+            throw Exception("Disk name can contain only alphanumeric and '_' (" + disk_name + ")", ErrorCodes::EXCESSIVE_ELEMENT_IN_CONFIG);
+
+        auto disk_config_prefix = config_prefix + "." + disk_name;
+        UInt64 keep_free_space_bytes = config.getUInt64(disk_config_prefix + ".keep_free_space_bytes", 0);
+        String path;
+        if (config.has(disk_config_prefix + ".path"))
+            path = config.getString(disk_config_prefix + ".path");
+
+        if (disk_name == default_disk_name)
+        {
+            has_default_disk = true;
+            if (!path.empty())
+                throw Exception("It is not possible to specify default disk path", ErrorCodes::UNKNOWN_ELEMENT_IN_CONFIG);
```

Comment:
Better to make exception more helpful - default disk path should be provided in <path> not it <storage_configuration>